(WIP) Add maxrequestbody directive

Ngo-The-Trung opened the pull request:
@mholt I'm looking for opinions on this implementation before I proceed further

[Issue](https://github.com/mholt/caddy/issues/1027)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/mholt/caddy?pullRequest=1163) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/mholt/caddy?pullRequest=1163) before we can accept your contribution.<br/>

mholt commented on caddyhttp/httpserver/siteconfig.go:
```diff
@@ -30,6 +30,14 @@ type SiteConfig struct {
 	// for a request.
 	HiddenFiles []string
+
+	// Max amount of bytes a request can send on a given path
```
"Max body size for requests to their respective paths"

mholt commented on caddyhttp/httpserver/siteconfig.go:
```diff
@@ -30,6 +30,14 @@ type SiteConfig struct {
+	MaxRequestBodySizes []PathLimit
+}
+
+type PathLimit struct {
```
This'll need some godoc of course

elcore replied:
Yes, please add some godoc.

mholt commented on caddyhttp/httpserver/server.go:
```diff
@@ -406,3 +417,14 @@ func WriteTextResponse(w http.ResponseWriter, status int, body string) {
+// Returns a boolean - whether the given error is a common error, and a int -
+// the suggested HTTP status code to return to the user
+func IsCommonError(err error) (bool, int) {
+	if err.Error() == "http: request body too large" {
```
Reading the string content of an error value is a little scary -- I like to avoid this where possible. It could change any time.

After looking into MaxBytesReader, I see it [makes use of an unexported error value](https://github.com/golang/go/blob/89c4cbd7acc36cf627746f379f0cc002b5b60383/src/net/http/request.go#L982), hence the need for this function which has to read the error string...

Finding https://github.com/golang/go/issues/8479, it seems like the Go team isn't interested in having the error value exported and reused, but the implementation of MaxBytesReader is trivial; we could just borrow from it and change it to meet our needs by using a pre-defined error value. Also it would be useful to mention how big the read body was, just so it's easier to read in the logs.

However, we should decide if we want to centralize the use of a MaxBodyReader like this, or if each middleware that reads the response body should use it instead. Either way, they'll have to check for the kind of error value returned and, if an EntityTooLargeError, they return a special status code in that case.

tw4452852 replied:
Agree with @mholt , Use the type assertion instead.

mholt commented on caddyhttp/maxrequestbody/maxrequestbody.go:
```diff
@@ -0,0 +1,59 @@
+	"github.com/mholt/caddy/caddyhttp/httpserver"
+)
+
+var SIZE_RE = regexp.MustCompile("[1-9][0-9]*[kmKM]?")
```
To prevent maintenance headaches in the future, can we not use a regular expression? :) This is a simple job for parsing. If it's a positive integer, use it as bytes. If it's a string suffixed with any case variation of KB, MB, (or GB, I suppose), split it out and parse the non-negative number before it.

Ngo-The-Trung replied:
Okay

tw4452852 commented on caddyhttp/maxrequestbody/maxrequestbody.go:
```diff
@@ -0,0 +1,59 @@
+	if size, err := strconv.ParseInt(arg, 10, 64); err != nil {
+		return c.ArgErr()
+	} else {
+		config.MaxRequestBodySizes = append(config.MaxRequestBodySizes, httpserver.PathLimit{path, size * mult})
```
Should we check the possible path duplication?

Ngo-The-Trung replied:
I've been meaning to ask this - do we:
- clean up the path (add slashes)
- sort the paths (so longer/more specific paths are checked first)

tw4452852 replied:
I think we could clean up the path here and if the path exists, overwrite.
Also sorting paths for specific path matching above makes sense.

mholt replied:
Hmm, I thought there was a facility in the `httpserver` package to help with the longest-path-matching thing, but maybe it was only the `proxy` middleware that got that blessed feature. (Man, I still really think we need to totally standardize the code to match requests. But that's a different topic.) Check out the proxy middleware maybe...

tw4452852 commented on caddyhttp/httpserver/server.go:
```diff
@@ -265,6 +265,17 @@ func (s *Server) serveHTTP(w http.ResponseWriter, r *http.Request) (int, error)
+	// .Read() or a similar method on the request body
+	for _, pathlimit := range vhost.MaxRequestBodySizes {
+		if Path(r.URL.Path).Matches(pathlimit.Path) {
+			r.Body = http.MaxBytesReader(w, r.Body, pathlimit.Limit)
```
`r.Body` may be `nil`.

mholt replied:
I've heard this debated before, and so I'm not sure if that's true -- but I definitely would not oppose to a nil check here!

mholt commented on caddyhttp/httpserver/plugin.go:
```diff
@@ -437,6 +437,7 @@ var directives = []string{
 	"hugo",        // github.com/hacdias/caddy-hugo
 	"mailout",     // github.com/SchumacherFM/mailout
 	"awslambda",   // github.com/coopernurse/caddy-awslambda
+	"maxrequestbody",
```
Also I wonder if it should be called `max_req_body` or something a little shorter (and readable).

Ngo-The-Trung replied:
Doesn't seem like anything other directives use underscore at the moment.

Also I will leave renaming last.

Ngo-The-Trung commented:
Updated. Addressed the comments above (except for the directive name, that comes later)

- Copied MaxBytesReader from the Go std here and use type checking for error
- Updated parsing to allow multiple paths (such as `maxrequestbody / 100MB /index 200GB`)
- Allow sizes of B, KB, MB, GB (optional B, case insensitive) 
- Parse sorts by path lengths
- Add some parsing tests

I don't understand the CI failures.

tw4452852 commented on caddyhttp/maxrequestbody/maxrequestbody_test.go:
```diff
@@ -0,0 +1,83 @@
+			t.Errorf("Expecting no error for case %v but encountered %v", caseNum, err)
+		}
+
+		if len(c.expected) != len(output) {
```
Maybe `reflect.DeepEqual` helps.

tw4452852 commented on caddyhttp/maxrequestbody/maxrequestbody_test.go:
```diff
@@ -0,0 +1,83 @@
+	for caseNum, c := range cases {
+		output := append([]PathLimit{}, c.arguments...)
+		SortPathLimits(output)
+		if len(c.expected) != len(output) {
```
Ditto.

tw4452852 commented on caddyhttp/maxrequestbody/maxrequestbody_test.go:
```diff
@@ -0,0 +1,83 @@
+import (
+	"testing"
+
+	. "github.com/mholt/caddy/caddyhttp/httpserver"
```
Explicit import may be better.

Ngo-The-Trung replied:
Not easy. 100 * KB is more readable than say 100 * validUnits["KB"] or 100 * validUnits[1], depending on whether I'm making it a map or a list.

Also I traverse the units in order in the parser code so if I were to rewrite validUnits to a map I would need to make an additional variable to store the order of the keys.

I would keep the constants.

Ngo-The-Trung replied:
Edit: I meant to reply to the `validUnits` comment

tw4452852 commented on caddyhttp/maxrequestbody/maxrequestbody_test.go:
```diff
@@ -0,0 +1,83 @@
+	. "github.com/mholt/caddy/caddyhttp/httpserver"
+)
+
+const (
```
Could these be combined with `validUnits` in `maxrequestbody.go` ?

tw4452852 commented on caddyhttp/maxrequestbody/maxrequestbody.go:
```diff
@@ -0,0 +1,152 @@
+	SortPathLimits(pathLimit)
+
+	config.MaxRequestBodySizes = pathLimit
+	fmt.Println(pathLimit)
```
Debug code?

Ngo-The-Trung replied:
Forgot to remove

tw4452852 commented on caddyhttp/maxrequestbody/maxrequestbody.go:
```diff
@@ -0,0 +1,152 @@
+	// Arguments appear in pairs of "<path> <limit>"
+	// If there's an argument that's not paired, it's assumed to be
+	// the limit used for all remaining paths ("/")
+	if len(args)%2 == 1 {
```
This could be wrong. Suppose we have this sequence: `/a 100KB 200KB`  or `/a 100KB /b`

Ngo-The-Trung replied:
Well I think everything should have a pair, with the only exception being the first argument.

The 2 cases you mentioned would fail due to the ParseInt below, also I have no idea how to interpret them.

tw4452852 replied:
Hmm, Is there any reason for this exception?

Ngo-The-Trung replied:
For convenience

`maxrequestbody 200MB` is understood as "limit everything (/) to 200MB"

`maxrequestbody 200MB /upload 500GB` is understood as "limit upload to 500GB, and limit everythign else to 200MB"

mholt replied:
I think this is a little confusing... A more Caddy-like way of doing pairs would be a "table" format. How about these valid forms instead:

```
maxrequestbody 200MB
maxrequestbody /upload 500GB
maxrequestbody {
    /ten 10MB
    /twenty 20MB
}
```

I think this way is much clearer.

tw4452852 commented on caddyhttp/httpserver/server.go:
```diff
@@ -406,3 +488,14 @@ func WriteTextResponse(w http.ResponseWriter, status int, body string) {
+	if _, ok := err.(MaxBytesExceeded); ok {
+		return true, http.StatusRequestEntityTooLarge
+	}
+	return false, -1
```
`-1` maybe a invalid HTTP status.

Ngo-The-Trung replied:
My intention is that the status code is advisory only, and is only applicable when the returned bool is `true`. It should be discarded if it is `false`.

tw4452852 replied:
Make it clear in comment.

tw4452852 commented on caddyhttp/httpserver/server.go:
```diff
@@ -392,6 +406,74 @@ func (ln tcpKeepAliveListener) File() (*os.File, error) {
 
+// MaxBytesExceeded is the error type returned by MaxBytesReader
+// when the request body exceeds the limit imposed
+type MaxBytesExceeded int
```
Is there any reason for using a `int` here ?

Ngo-The-Trung replied:
No. Could be anything. Should I use `interface{}` instead?

tw4452852 replied:
Maybe `struct{}` better.

tw4452852 commented:
@Ngo-The-Trung 
Also for CI issue, you should check `TestStandardPlugins`, I think you will find the answer.

elcore commented on caddyhttp/maxrequestbody/maxrequestbody.go:
```diff
@@ -0,0 +1,150 @@
+			return pathLimit, errors.New("Parse failed")
+		} else {
+			pathLimit = addPathLimit(pathLimit, path, size)
+		}
```
You do not need to use else here

elcore commented on caddyhttp/maxrequestbody/maxrequestbody.go:
```diff
@@ -0,0 +1,150 @@
+				return -1
+			} else {
+				return size * unit.multiplier
+			}
```
You do not need to use else here

elcore commented on caddyhttp/maxrequestbody/maxrequestbody.go:
```diff
@@ -0,0 +1,150 @@
+	return s.by(&s.pathLimits[i], &s.pathLimits[j])
+}
+
+func LengthDescending(p1, p2 *httpserver.PathLimit) bool {
```
We need some godoc here

elcore commented on caddyhttp/maxrequestbody/maxrequestbody_test.go:
```diff
@@ -0,0 +1,118 @@
+				httpserver.PathLimit{Path: "/home", Limit: 12345 * MB},
+			},
+			hasError: false,
+		},
```
Please use gofmt

Ngo-The-Trung replied:
My gofmt seems to be ok with this for some reason

Ngo-The-Trung replied:
Oh gofmt wasn't the reason. You guys are using simplify mode (-s).

I'll update this.

elcore replied:
Well, it´s part of gofmt 😉 -- Our code should be simple 😄

Ngo-The-Trung commented:
Well at least the CI is green now

elcore commented:
> Well at least the CI is green now

@Ngo-The-Trung mission accomplished 😄

Ngo-The-Trung commented:
Pinging @mholt @tw4452852 for review 

Not sure how to test the limiting itself.

tw4452852 commented on caddyhttp/maxrequestbody/maxrequestbody.go:
```diff
@@ -0,0 +1,150 @@
+		path = "/" + path
+	}
+
+	// Avoid duplicates
```
I think it's better to overwrite the exist one.

tw4452852 commented on caddyhttp/maxrequestbody/maxrequestbody_test.go:
```diff
@@ -0,0 +1,118 @@
+			t.Errorf("Expecting no error for case %v but encountered %v", caseNum, err)
+		}
+
+		if reflect.DeepEqual(c.expected, len(output)) {
```
Should be: `!reflect.DeepEqual(c.expected, output)`

Ngo-The-Trung replied:
Oh lol the 2 wrongs make it right

tw4452852 commented on caddyhttp/maxrequestbody/maxrequestbody_test.go:
```diff
@@ -0,0 +1,118 @@
+	for caseNum, c := range cases {
+		output := append([]httpserver.PathLimit{}, c.arguments...)
+		SortPathLimits(output)
+		if reflect.DeepEqual(c.expected, len(output)) {
```
Ditto.

Ngo-The-Trung commented:
Updated

mholt commented:
Making progress! :) I'll look at it more when I'm back from my trip.

mholt commented on caddyhttp/httpserver/server.go:
```diff
@@ -406,3 +488,15 @@ func WriteTextResponse(w http.ResponseWriter, status int, body string) {
+// Returns a boolean - whether the given error is a common error, and an int -
+// the advisory HTTP status code to return to the user (or -1 if the given error
+// is not a common error)
+func IsCommonError(err error) (bool, int) {
```
I still don't quite feel good about this function. It's only used by MaxBytesReader, I think error handling should not be here. It should probably remain in the handlers that use the MaxBytesReader.

Ngo-The-Trung replied:
Ok

Ngo-The-Trung commented:
Updated

mholt commented on caddyhttp/maxrequestbody/maxrequestbody.go:
```diff
@@ -0,0 +1,151 @@
+			pathLimit[i].Limit = limit
+			return pathLimit
+		}
+	}
```
Nice, glad you're thinking of stuff like this. :+1:

mholt commented on caddyhttp/proxy/proxy.go:
```diff
@@ -174,6 +174,10 @@ func (p Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error) {
 
+		if _, ok := backendEerr.(httpserver.MaxBytesExceeded); ok {
+			return http.StatusRequestEntityTooLarge, backendErr
+		}
```
_NICE._ This is what I was looking for. :tada:

mholt commented:
Cool, I'm almost done with a final review -- looking pretty good so far (I might massage a few things after merging, no big deal). Just need a few more minutes. :) Perhaps meanwhile @tw4452852 and @elcore would be able to update their reviews or approve the changes.

tw4452852 commented on caddyhttp/maxrequestbody/maxrequestbody_test.go:
```diff
@@ -0,0 +1,127 @@
+	MB = 1024 * 1024
+	GB = 1024 * 1024 * 1024
+)
+
```
Add `TestSetupMaxRequestBody` ? (just test the all kinds of syntax formats)

tw4452852 commented on caddyhttp/maxrequestbody/maxrequestbody_test.go:
```diff
@@ -0,0 +1,127 @@
+		{arguments: []pathLimitUnparsed{{"/", "200LB"}}, expected: []httpserver.PathLimit{}, hasError: true},
+		{arguments: []pathLimitUnparsed{{"/", "path:999MB"}}, expected: []httpserver.PathLimit{}, hasError: true},
+		{arguments: []pathLimitUnparsed{{"/", "1_234_567"}}, expected: []httpserver.PathLimit{}, hasError: true},
+
```
Add test with zero size.

Ngo-The-Trung commented:
@tw4452852 done

tw4452852 commented on caddyhttp/maxrequestbody/maxrequestbody.go:
```diff
@@ -0,0 +1,178 @@
+		// Format: { <path> <limit> ... }
+		for c.NextBlock() {
+			path := c.Val()
+			if !c.NextBlock() {
```
I think we should use `NextArg` here.

Ngo-The-Trung replied:
Done

mholt commented:
@tw4452852 When you are satisfied, approve and merge away!

Ngo-The-Trung commented:
@mholt moved to below "bind"

tw4452852 commented:
LGTM. Thanks @Ngo-The-Trung !

mholt commented:
@Ngo-The-Trung Thank you for your work on this!

Would you like to be a collaborator? I'll send you the invite and you may accept it if you want. This means you'd be able to contribute to Caddy without having to fork the project (you could commit to branches) and you can help by reviewing PRs, etc. Really appreciate your efforts. :)
